Improved display behavior

DionTimmermann opened the pull request:
At the moment, this is not a real pull-request, more an info and request for comments.

To help me debug some code, I wanted a nicer display of variables in the command line. The code has not been fully tested and currelty is only implemented for LinProp variables.

Display of a scalar in the command line:
```
>> LinProp(rand, rand)

ans = 

    LinProp:
     (0.6171 ± 0.2653)

Show Budget, Methods.
```

The words 'Buget' and 'Methods' are links. Clicking on the 'Budget' link prints the following:
```
  Budget of ans:
    Description                                       Component    Percentage
    Unknown Id                                           0.2653    100.00%
      B5-80-29-AB-36-A1-4C-8C-8D-8D-B9-B8-6F-EB-0A-07    0.2653    100.00%

```

Matricies are printed column-wise:
```
>> LinProp(rand(4, 6))

ans = 

    4×6 LinProp:

  Columns 1 through 4
     (0.8244 ± 0)     (0.5841 ± 0)     ( 0.8178 ± 0)     (0.4253 ± 0)
     (0.9827 ± 0)     (0.1078 ± 0)     ( 0.2607 ± 0)     (0.3127 ± 0)
     (0.7302 ± 0)     (0.9063 ± 0)     ( 0.5944 ± 0)     (0.1615 ± 0)
     (0.3439 ± 0)     (0.8797 ± 0)     (0.02251 ± 0)     (0.1788 ± 0)

  Columns 5 through 6
     ( 0.4229 ± 0)     (0.6959 ± 0)
     (0.09423 ± 0)     (0.6999 ± 0)
     ( 0.5985 ± 0)     (0.6385 ± 0)
     ( 0.4709 ± 0)     (0.0336 ± 0)

Show Methods.
```

This also works with complex matricies of more than 2 dimensions:
```
>> LinProp(rand(3, 2, 2))+1i*LinProp(rand(3, 2, 2))

ans = 

    3×2×2 LinProp:

(:,:,1) =

     ( 0.7788 ± 0) + (0.6377 ± 0)i     (0.2665 ± 0) + (0.6761 ± 0)i
     ( 0.4235 ± 0) + (0.9577 ± 0)i     (0.1537 ± 0) + (0.2891 ± 0)i
     (0.09082 ± 0) + (0.2407 ± 0)i     ( 0.281 ± 0) + (0.6718 ± 0)i


(:,:,2) =

     (0.4401 ± 0) + ( 0.6951 ± 0)i     (0.8754 ± 0) + ( 0.224 ± 0)i
     (0.5271 ± 0) + (0.06799 ± 0)i     (0.5181 ± 0) + (0.6678 ± 0)i
     (0.4574 ± 0) + ( 0.2548 ± 0)i     (0.9436 ± 0) + (0.8444 ± 0)i

Show Methods.
```

wollmich commented:
Merging brain 'main' into this pull request didn't work as expected. Can you please help to merge?

DionTimmermann commented:
Sorry, I had accidentally caused a conflict. I was working on this branch anyways and will update it shortly and merge main.

DionTimmermann commented:
OK, I have mergerd main and cleaned up the code a bit. I think I am now happy with this branch. 

Since this pull-request affects the display behaviour, I did not see any way to automate it's testing.

wollmich commented on @LinProp/LinProp.m:
```diff
@@ -2089,17 +2250,17 @@ function xml_file(x, filepath)
                 temp = temp*s(i2); 
             end
         end
-        function TF = IsComplexNet(x)
```
Why are there changes in the `IsComplexNet` and in the `IsArrayNet` methods?

wollmich commented on @LinProp/LinProp.m:
```diff
@@ -2188,4 +2349,4 @@ function xml_file(x, filepath)
         end
     end 
-end
+end
```
A newline is missing at the end.

wollmich commented on _Copy_LinProp_to_DistProp_and_MCProp.cmd:
```diff
@@ -1,4 +0,0 @@
-powershell -Command "(gc @LinProp\LinProp.m) -replace 'LinProp', 'DistProp' | Out-File -encoding default @DistProp\DistProp.m"
```
Why did you removed this file?

wollmich commented on @LinProp/LinProp.m:
```diff
@@ -392,44 +341,38 @@ function display(obj)
-                        s(end+1:nargout) = 1;
-                    elseif nargout < length(s)
-                        s = [s(1:nargout-1) prod(s(nargout:end))];
+            switch (numel(varargin))
```
Why are there changes in the size function?

DionTimmermann commented:
I am sorry, it seems my prior merge from main had some unintended regressions. It seems I still sometimes struggle with GIT. None of the changes you marked in the review were inteded.

wollmich commented:
When I click on `Contributions` then I get an error using MATLAB R2018a, see the following example:

```
>> a = LinProp(3, 0.3)

a = 

    LinProp:
     (3 ± 0.3)

Show in Format long, Contributions, Methods.
>> b = LinProp(4, 0.4)

b = 

    LinProp:
     (4 ± 0.4)

Show in Format long, Contributions, Methods.
>> c = sqrt(a.^2 + b.^2)

c = 

    LinProp:
     (5 ± 0.3672)

Show in Format long, Contributions, Methods.

  Variable c has 2 uncertainty contributions.
  Top level contributions are:

    Description                                         Component    Percentage
    Unknown Id
      0C-F4-C5-05-54-12-4C-CE-99-59-09-6D-10-34-AC-8D        0.32     75.96%
      63-2B-74-C4-B7-00-4C-D2-85-4B-28-08-93-D7-15-CF        0.18     24.04%

The class LinProp has no Constant property or Static method named 'getMethodLink'.

Error in LinProp/displayContributions (line 112)
    fprintf('%s.\n', LinProp.getMethodLink(methodCall, 'Open full budget', varName,
    class(obj)));
```

DionTimmermann commented:
I cannot reproduce this problem, neither with Matlab 2021a nor 2018a. `getMethodLink()` used to be a static method, but I changed it to a local function. Could it be that you did not pull the newest version of [@LinProp/displayContributions.m](https://github.com/wollmich/metas-unclib-matlab-wrapper/pull/46/files#diff-212d01da4701f77bc17c7b203babd0cad4c29c6605523d505386791f221a2acc)?

In your example, the Ids of both numbers are shown. In my example below they are simply labeled as 'Unknown'. This could either be due to an older version of `displayContributions()`, or because the test `if exist('LoadVNATools', 'file')` fails on your system, i.e. the VnaTools matlab folder is not on your matlab path.

When I run your example, I get: 
```
a = 

    LinProp:
     (3 ± 0.3)

Show in Format long, Contributions, Methods.

b = 

    LinProp:
     (4 ± 0.4)

Show in Format long, Contributions, Methods.

c = 

    LinProp:
     (5 ± 0.3672)

Show in Format long, Contributions, Methods.

  Variable c has 2 uncertainty contributions.
  Top level contributions are:

    Description             Component    Percentage
    Unknown                      0.32     75.96%
    Unknown                      0.18     24.04%

Open full budget.
```

wollmich commented:
> Could it be that you did not pull the newest version of [@LinProp/displayContributions.m](https://github.com/wollmich/metas-unclib-matlab-wrapper/pull/46/files#diff-212d01da4701f77bc17c7b203babd0cad4c29c6605523d505386791f221a2acc)?

Yes I didn't fetched the 3 last commits. Sorry about that. Now it works as well on my side.

wollmich commented:
> or because the test if exist('LoadVNATools', 'file') fails on your system, i.e. the VnaTools matlab folder is not on your matlab path.

What's the logical behind that? I would prefer to see the Unknown IDs in both cases.

DionTimmermann commented:
That was a side effect form the `Metas.Vna.Data.UncIdDefs.GetInfluenceInfo2` decoder. Without it, the tree generated by `ComputeTreeUncBudget` contains the IDs, with the decoder, the tree does not contain the IDs.

I have pushed a commit that adds in the IDs for contributions labeled 'Unknown'. Is that a good solution? I am hesitant to show the IDs for LinProps that have a description that is not 'Unknown'- as that would likely result in a very wide table. Of cause, I could also add another link 'Show Contributions with IDs'.

The IDs curretly are formatted like GUIDs, because I was able to just copy that code from somewhere else. Would you like me to change the code so the IDs are formatted in 8bit-pairs, as in the budget viewer? (This will result in more dashes and thus wider tables.)

```
>> a = LinProp(3, 0.3, 'My new variable')

a = 

    LinProp:
     (3 ± 0.3)

Show in Format long, Contributions, Methods.
>> b = LinProp(4, 0.4)

b = 

    LinProp:
     (4 ± 0.4)

Show in Format long, Contributions, Methods.
>> c = sqrt(a.^2 + b.^2)

c = 

    LinProp:
     (5 ± 0.3672)

Show in Format long, Contributions, Methods.

  Variable c has 2 uncertainty contributions.
  Top level contributions are:

    Description                                      Component    Percentage
    My new variable                                       0.18     24.04%
    Unknown (202d0ad0-2a37-47a4-bed2-b7c0c5d6ed77)        0.32     75.96%

Open full budget.
```

Using the decoder is useful for working and testing data from VNA tools:
```
>> u = prf.support.uncertaintyContribution('VnaLinearity', ...
    0, 1, ...
    'VnaId', vnaId, ...
    'ReceiverPortNumber', 2, ...
    'PowerIndex', 1e6*(0.02+1000), ...
    'isPhase', false...
)

u = 

    LinProp:
     (0 ± 1)

Show in Format long, Contributions, Methods.

  Variable u has 1 uncertainty contribution.
  Top level contributions are:

    Description                  Component    Percentage
    VNA Linearity → S-Parameter
      @ 0.020000 dB, Port: 2, Mag        1    100.00%

Open full budget.
```

wollmich commented:
> That was a side effect form the `Metas.Vna.Data.UncIdDefs.GetInfluenceInfo2` decoder. Without it, the tree generated by `ComputeTreeUncBudget` contains the IDs, with the decoder, the tree does not contain the IDs.

Yes you're right. I'll fix that behavior on my side in C# either in `Metas.UncLib.LinProp.UncBudget.ComputeFlatUncBudget` or in `Metas.Vna.Data.UncIdDefs.GetInfluenceInfo2`. `ComputeTreeUncBudget` calls compute `ComputeFlatUncBudget`.

So there is no need that you try to decode the ID on your side. So commit https://github.com/wollmich/metas-unclib-matlab-wrapper/pull/46/commits/287cdf2f7e0a590fe89d398760ce0f596d735fd2 would not be needed.

wollmich commented:
I've fix this bug in `Metas.Vna.Data.UncIdDefs.GetInfluenceInfo2`. This will be part of METAS VNA Tools V2.5.4.

wollmich commented on @LinProp/displayContributions.m:
```diff
@@ -0,0 +1,125 @@
+    end
+    
+    if exist('LoadVNATools', 'file')
+        LoadVNATools();
```
This could yield into problems when the version of METAS UncLib and METAS VNA Tools are not the same. E.g.: `Metas.UncLib.Core.dll` of METAS VNA Tools could be older or newer than the already loaded assembly from METAS UncLib.

On our side this handled like that: 
- The first UncLib object is created without calling `LoadVNATools.m` first. In that case the UncLib assemblies are loaded from the METAS UncLib location.
- `LoadVNATools.m` is called before creating the first UncLib oject. In that case the UncLib assemblies are loaded from the METAS VNA Tools location and the `Metas.Vna.Data.UncIdDefs.GetInfluenceInfo2` exists.

See as well how I handle this in `unc_budget.m`.

DionTimmermann replied:
Oh, I never thought of that... Thank you for pointing this out.

I have pushed a commit that changes these lines to a try-catch, similar to the way you handled it in `unc_budget.m`.

wollmich commented:
I've one more comment. `LinProp(randn(2,9))` generates the following output using this pull request:

```
ans = 

    2×9 LinProp:
  Columns 1 through 5
     (0.5377 ± 0)    -( 2.259 ± 0)     (0.3188 ± 0)    -(0.4336 ± 0)     (3.578 ± 0)
     ( 1.834 ± 0)     (0.8622 ± 0)    -( 1.308 ± 0)     (0.3426 ± 0)     (2.769 ± 0)

  Columns 6 through 9
    -( 1.35 ± 0)     ( 0.7254 ± 0)     (0.7147 ± 0)    -(0.1241 ± 0)
     (3.035 ± 0)    -(0.06305 ± 0)    -( 0.205 ± 0)     (  1.49 ± 0)

Show in Format long, Methods.
```

Before the following output has been generated:

```
ans.value = 
 
  Columns 1 through 8

    0.5377   -2.2588    0.3188   -0.4336    3.5784   -1.3499    0.7254    0.7147
    1.8339    0.8622   -1.3077    0.3426    2.7694    3.0349   -0.0631   -0.2050

  Column 9

   -0.1241
    1.4897

ans.standard_unc = 
 
     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0     0
```

What I liked better in the old display:

- The decimal point `.` was always at the same position
- The number of digits after the decimal points was the same for the whole matrix.

DionTimmermann commented:
I am afraid there is no easy solution.

Both points you mention are side effects of the new display using `printf('%g', data)`, which sometimes uses an exponent, and sometimes not, depending on the magnitude of the number. It also removes trailing zeros. The old display used something similar to `printf('%f', data)`. I prefer `%g` as the value and stdunc often have different orders of magnitude, as in the example below. 
```
example = 

    2×2 LinProp:
     (    0 ± 103.1) + (0 ± 103.1)i     (1e+07 ± 563.2) + (0 ± 512.1)i
     (1e+07 ±     0) + (0 ±     0)i     (    0 ± 103.1) + (0 ± 103.1)i
```
Printing the same data in style more similar to the old display would look as follows:
```
example = 

    2×2 LinProp:

  1.0e+07 *
    (0.0000 ± 0.0000) + (0.0000 ± 0.0000)i     (1.0000 ± 0.0001) + (0.0000 ± 0.0001)i
    (1.0000 ± 0.0000) + (0.0000 ± 0.0000)i     (0.0000 ± 0.0000) + (0.0000 ± 0.0000)i
```
There, we would extract the largest exponent and effectivly print every number as `printf('%f', data/largestExponent)`. The drawback of this approach is that the stdunc is practically gone and could only be seen in long format:
```
example = 

    2×2 LinProp:

  1.0e+07 *

  Column 1
     (0.000000000000000 ± 0.000010308676339) + (0.000000000000000 ± 0.000010308676339)i
     (1.000000000000000 ± 0.000000000000000) + (0.000000000000000 ± 0.000000000000000)i

  Column 2
     (1.000000000000000 ± 0.000056320612591) + (0.000000000000000 ± 0.000051206227499)i
     (0.000000000000000 ± 0.000010308676339) + (0.000000000000000 ± 0.000010308676339)i
```

I think it is important to not remove the `%g` formatting and see two ways to solve this issue:
1) Make display respect the `format shortg`, `format shortf`,  `format longg`, and `format longf`,  settings. Then, the user could choose how to display unc numbers. This would require some rework of the code to extract and print the largest exponent. We would also have to decide if `format short` were be equal to `format shortf` or `format shortg`...
2) Always use `%g` as I have currently implemented it, but manually fix the display using string operations. (Shifting the numbers so the `.` align and adding tailing zeros so all numbers have the same number of digits.)

Which approach do you prefer? (Technically, these two options are not mutually exclusive.)

wollmich commented:
Thank you for clarifying.

>  Which approach do you prefer? (Technically, these two options are not mutually exclusive.)

None of them. Keep it as you've it right now.

I would like to have the possibility to switch between my original display format and your new display format. What do you think about it?

> There, we would extract the largest exponent and effectivly print every number as `printf('%f', data/largestExponent)`. The drawback of this approach is that the stdunc is practically gone and could only be seen in long format:

I don't have that problem in my original display format, see the following example:

```
>> LinProp(1e9.*randn(2, 2), 1e-9*eye(4))
 
ans.value = 
 
   1.0e+09 *

    0.7147   -0.1241
   -0.2050    1.4897

ans.standard_unc = 
 
   1.0e-04 *

    0.3162    0.3162
    0.3162    0.3162
```

DionTimmermann commented:
> I would like to have the possibility to switch between my original display format and your new display format. What do you think about it?

I think that would be a way better solution than the two I suggested above. 

I have made almost all changes, but am still working on some parts of the formatting. How would you like to be able to switch between the formats? With the links below the output? It seems as if I will need add a global variable anyways to implement the switch between the formats, so that could also be documented for the user. However, I still need to finish the work on that.

wollmich commented:
> How would you like to be able to switch between the formats?

An additional global variable, like `DistPropGlobalMaxLevel` or `MCPropGlobalN(n)`. The name could be for example `UncLibImprovedDisplay` and the argument 1 turns it on and 0 turns it off.  For the moment I would prefer to keep the old display behavior as default. I like as well the idea with the links.

> With the links below the output?

That would be very nice.

wollmich commented:
This pull request needs to be updated after the following pull requests:

- https://github.com/wollmich/metas-unclib-matlab-wrapper/pull/58
- https://github.com/wollmich/metas-unclib-matlab-wrapper/pull/60
- https://github.com/wollmich/metas-unclib-matlab-wrapper/pull/63
- https://github.com/wollmich/metas-unclib-matlab-wrapper/pull/64

DionTimmermann commented:
I now have incorporated the all the changes from the main branch, which has significantly simplified the code. 

There now exists a global property `UncLibMatrixDisplay`, which can have the values `separate` or `combined`. It can be changed using the links in the comand line:
![grafik](https://user-images.githubusercontent.com/14871355/173302474-c1ac8dcb-5622-4bfc-8971-90b223d083fc.png)

I have have not fully tested the changes, but will use and tesst this branch in the next week(s).

DionTimmermann commented:
The changes yesterday resulted in some matricies to not have the plus/minus signs aligned. I have now fixed that. I have also fixed a but in #67.

Now matricies are alinged again (and the decimal dots are also aligned):
![grafik](https://user-images.githubusercontent.com/14871355/173521367-0a765d35-33c0-488e-a513-2ef362bf9153.png)

DionTimmermann commented:
I have now updated this branch to incorporate all changes in the main branch. 

Overall, I am happy with this branch. However there still are some edge cases that I would like to improve and I have spotted one error in the last weeks that prevented the printing of a maxtrix. Unfortunately, I was not able to reproduce this error. Therefore, I think this pull request still is not ready for review. (I will only be able to work on this branch again at the end of October 2022.)

wollmich commented:
> Therefore, I think this pull request still is not ready for review. (I will only be able to work on this branch again at the end of October 2022.)

Absolutely no hurry. We keep this pull request marked as draft. Once you're ready please let me know.

DionTimmermann commented:
I have now closed the last bug that I was aware of. I have now used this code for about 4 months without any issues. Still, I would reccomend extensive tests, as I surely did not use any imaginable type of variable/matrix.

Known issues:
- For very large matricies, the output in the combined format takes longer than I would like it to, but I doubt that optimizing that is worth it. 
- The way to change the output formating (combined/separate uncertainties) is not documented well.

DionTimmermann commented:
This branch is affected by #83, as it uses `ind2subs` multiple times.
